crates/core/src/protocols/sumcheck/v3/bivariate_mlecheck.rs

```diff
@@ -0,0 +1,535 @@
+use binius_utils::bail;
+
+use super::bivariate_product::{PhaseState, SumcheckMultilinear};
+use crate::{
+	composition::{BivariateProduct, IndexComposition},
+	protocols::sumcheck::{
+		CompositeSumClaim, EqIndSumcheckClaim, Error, RoundCoeffs, prove::SumcheckProver,
+	},
+};
+
+/// MLEcheck prover implementation for the special case of bivariate product compositions over
+/// large-field multilinears.
+///
+/// This implements the [`SumcheckProver`] interface. The implementation uses a [`ComputeLayer`]
+/// instance for expensive operations and the input multilinears are provided as device memory
+/// slices.
+pub struct BivariateMLEcheckProver<'a, 'alloc, F: Field, Hal: ComputeLayer<F>> {
+	hal: &'a Hal,
+	dev_alloc: &'a BumpAllocator<'alloc, F, Hal::DevMem>,
+	host_alloc: &'a HostBumpAllocator<'a, F>,
+	n_vars_initial: usize,
+	n_vars_remaining: usize,
+	multilins: Vec<SumcheckMultilinear<'a, F, Hal::DevMem>>,
+	compositions: Vec<IndexComposition<BivariateProduct, 2>>,
+	last_coeffs_or_sums: PhaseState<F>,
+	eq_ind_prefix_eval: F,
+	// Wrapping it in Option is a temporary workaround for the lifetime problem
+	eq_ind_partial_evals: Option<FSliceMut<'a, F, Hal>>,
+	eq_ind_challenges: Vec<F>,
+}
+
+impl<'a, 'alloc, F, Hal> BivariateMLEcheckProver<'a, 'alloc, F, Hal>
+where
+	F: TowerField,
+	Hal: ComputeLayer<F>,
+{
+	#[allow(clippy::too_many_arguments)]
+	pub fn new(
+		hal: &'a Hal,
+		dev_alloc: &'a BumpAllocator<'alloc, F, Hal::DevMem>,
+		host_alloc: &'a HostBumpAllocator<'a, F>,
+		claim: &EqIndSumcheckClaim<F, IndexComposition<BivariateProduct, 2>>,
+		multilins: Vec<FSlice<'a, F, Hal>>,
+		// Specify an existing tensor expansion for `eq_ind_challenges`. Avoids
+		// duplicate work.
+		eq_ind_partial_evals: FSlice<'a, F, Hal>,
+		eq_ind_challenges: Vec<F>,
+	) -> Result<Self, Error> {
+		if Hal::DevMem::MIN_SLICE_LEN != 1 {
+			todo!("support non-trivial minimum slice lengths");
+		}
+
+		let n_vars = claim.n_vars();
+
+		// Check shape of multilinear witness inputs.
+		assert_eq!(claim.n_multilinears(), multilins.len());
+		for multilin in &multilins {
+			if multilin.len() != 1 << n_vars {
+				bail!(Error::NumberOfVariablesMismatch);
+			}
+		}
+
+		// Wrap multilinear witness inputs as SumcheckMultilinears.
+		let multilins = multilins
+			.into_iter()
+			.map(SumcheckMultilinear::PreFold)
+			.collect();
+
+		let (compositions, sums) = claim
+			.eq_ind_composite_sums()
+			.iter()
+			.map(|CompositeSumClaim { composition, sum }| (composition.clone(), *sum))
+			.unzip();
+
+		// Only one value of the expanded equality indicator is used per each
+		// 1-variable subcube, thus it should be twice smaller.
+		let mut eq_ind_partial_evals_buffer = dev_alloc.alloc(1 << (n_vars - 1))?;
+		if eq_ind_partial_evals.len() != 1 << n_vars.saturating_sub(1) {
+			bail!(Error::IncorrectEqIndPartialEvalsSize);
+		}
+
+		hal.copy_d2d(eq_ind_partial_evals, &mut eq_ind_partial_evals_buffer)?;
```

Comment:
As follow-up: remove this copy and defer it to the fold phase. You can reuse the `SumcheckMultilinear` pattern with Prefold and Postfold.